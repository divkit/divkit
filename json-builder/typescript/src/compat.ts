import { DivLinearGradient } from './generated';

/**
 * @deprecated use DivLinearGradient instead.
 */
export class DivGradientBackground extends DivLinearGradient {}
