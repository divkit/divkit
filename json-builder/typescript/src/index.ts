export * from './compat';
export * from './div-utils';
export * from './expression';
export * from './generated';
export * from './helper';
export * from './safe-expression';
export * from './template';
export * from './template-helper';
export * from './template-helper-deps';
export * from './template-helper-memo';
export * from './template-helper-rewrite-names';
export * from './template-helper-rewrite-refs';
export * from './template-helper-version';
