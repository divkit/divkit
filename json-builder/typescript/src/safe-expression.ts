const expr = Symbol();

export interface DivExpression {
    [expr]: string;
}

export class SafeDivExpression implements DivExpression {
    public [expr]: string;

    public constructor(expression: string) {
        this[expr] = expression;
    }

    public toJSON(): string {
        return this[expr];
    }

    public toString(): string {
        return this[expr];
    }
}
