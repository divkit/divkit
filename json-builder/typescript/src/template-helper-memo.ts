import { ITemplates, TemplateBlock } from './template';
import { TemplateHelper } from './template-helper';

export interface ThelperWithMemo<T extends ITemplates> {
    thelper: TemplateHelper<T>;
    used: Set<string>;
}

/**
 * templateHelper with memoization of templates names for later
 * selection of used templates. It memoizes only explicitly used
 * names without tracking dependencies.
 * @param options
 * @param options.customName adds templates names mapping.
 * Instead of original template name adjust template's type to custom value
 * returned by customName. Can be used for templates versioning.
 */
export function thelperWithMemo<T extends ITemplates>(options?: {
    customName?: (x: string) => string;
}): ThelperWithMemo<T> {
    const used = new Set<string>();

    const proxyHandler: ProxyHandler<{ [k: string]: (o: unknown) => TemplateBlock<string> }> = {
        get(_, templateName: string) {
            const templateRenamed = options?.customName ? options.customName(templateName) : templateName;
            used.add(templateName);

            return (props: any) => ({ ...props, type: templateRenamed });
        },
    };

    const thelper = new Proxy({}, proxyHandler) as TemplateHelper<T>;

    return { used, thelper };
}
