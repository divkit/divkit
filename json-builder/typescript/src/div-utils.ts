import { DivFixedSize, DivMatchParentSize, DivSizeUnit, DivWrapContentSize } from './generated';
import { DivExpression } from './safe-expression';

export function fixed(value: number | DivExpression, unit: DivSizeUnit | undefined = undefined): DivFixedSize {
    if (unit === undefined) {
        return new DivFixedSize({ value: value });
    }
    return new DivFixedSize({ unit: unit, value: value });
}

export function matchParent(): DivMatchParentSize {
    return new DivMatchParentSize();
}

export function weighted(weight: number): DivMatchParentSize {
    return new DivMatchParentSize({ weight: weight });
}

export function wrapContent(): DivWrapContentSize {
    return new DivWrapContentSize();
}
