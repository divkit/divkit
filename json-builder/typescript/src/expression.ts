import { DivExpression, SafeDivExpression } from './safe-expression';

export function expression(expression: string): DivExpression {
    return new SafeDivExpression(expression);
}

const replacer: Record<string, string> = {
    '\\': '\\\\',
    '@{': '\\@{',
};

export function escapeExpression(str: string): string {
    return str.replace(/\\|(@{)/g, (full) => {
        return replacer[full];
    });
}
