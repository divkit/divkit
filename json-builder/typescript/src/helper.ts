import { ITemplates } from './template';
import { SafeDivExpression } from './safe-expression';

export type NonEmptyArray<T> = T[];

export type Exact<TBase, TExt extends TBase> = TExt extends unknown ? Exactly<TBase, TExt> : never;

type Exactly<TBase, TExt extends TBase> = {
    [K in keyof TExt]: K extends keyof TBase ? TExt[K] : never;
};

export type IntBoolean = 1 | 0 | true | false;

/**
 * DFS for a js object with callback on each leaf
 * @param tree js object
 * @param nodeAction callback for each leaf
 */
export function treeWalkDFS(tree: unknown, nodeAction: (x: unknown, path: string[]) => void): void {
    const stack: [unknown, string[]][] = [[tree, []]];

    while (stack.length) {
        const [node, path] = stack.pop() as [unknown, string[]];
        nodeAction(node, path);

        if (typeof node === 'object' && node !== null) {
            const list = Object.entries(node as { [k: string]: unknown }).map(
                ([key, obj]) => [obj, [...path, key]] as [unknown, string[]],
            );

            for (let i = list.length - 1; i >= 0; --i) {
                stack.push(list[i]);
            }
        }
    }
}

export function copyTemplates<T extends ITemplates>(templates: T): T {
    const copy: ITemplates = {};

    const placeHolder = (node: unknown): unknown => {
        if (!node || ['string', 'number', 'boolean'].includes(typeof node) || node instanceof SafeDivExpression) {
            return node;
        }

        if (Array.isArray(node)) {
            return [];
        }

        return Object.create(Object.getPrototypeOf(node));
    };

    const clone = (node: unknown, path: string[]): void => {
        let parentPointer: object = copy;

        path = [...path];
        while (path.length > 1) {
            parentPointer = parentPointer[path.shift() as string];
        }

        const cur = path[0];

        if (!cur) {
            return;
        }

        parentPointer[cur] = placeHolder(node);
    };

    treeWalkDFS(templates, clone);

    return copy as T;
}
