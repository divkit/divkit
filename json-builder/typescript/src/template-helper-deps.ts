import { Div } from './generated/Div';
import { treeWalkDFS } from './helper';
import { ITemplates, TemplateBlock } from './template';

/**
 * Search for a used templates list
 * @param template
 */
function findPlainDeps(template: Div): Set<string> {
    const deps = new Set<string>();

    treeWalkDFS(template, (node) => {
        if (node instanceof TemplateBlock) {
            deps.add(node.type);
        }
    });

    return deps;
}

const inProgress = 1;
const isFinished = 2;

export type TemplateResolvedAction<T> = (props: {
    name: string;
    template: Div;
    plainDeps: string[];
    depsResolved: { [k: string]: T };
}) => T;

/**
 * Templates walker. It resolves template dependencies on each step, and also can run a callback after that
 * @param templates
 * @param resolvedAction Callback for the template. It will be called after resolving dependencies
 * @param depsResolved Deps resolve cache, for example, for common templates
 */
export function runResolveDeps<T>(
    templates: ITemplates,
    resolvedAction: TemplateResolvedAction<T>,
    depsResolved: { [k: string]: T } = {},
): { [k: string]: T } {
    depsResolved = { ...depsResolved };
    const progress: { [k: string]: undefined | typeof inProgress | typeof isFinished } = {};

    for (const k of Object.keys(depsResolved)) {
        progress[k] = isFinished;
    }

    const stack = Object.keys(templates);

    while (stack.length) {
        const name = stack.pop() as string;

        if (progress[name] === isFinished) {
            continue;
        }

        const plainDeps = [...findPlainDeps(templates[name])];
        const unresolvedDeps = plainDeps.filter((dep) => !(progress[dep] === isFinished));
        const unsolvableDeps = unresolvedDeps.filter((depName) => !templates[depName]);

        if (unsolvableDeps.length !== 0) {
            throw new Error(
                `template '${name}' unsolvable dependencies: ${unsolvableDeps.map((b) => `'${b}'`).join(',')}`,
            );
        }

        if (unresolvedDeps.length === 0) {
            const template = templates[name];
            depsResolved[name] = resolvedAction({ name, template, plainDeps, depsResolved });
            progress[name] = isFinished;
        } else if (progress[name] === inProgress) {
            throw new Error(`template ${name}: cyclic depdendencies`);
        } else {
            progress[name] = inProgress;
            stack.push(name, ...unresolvedDeps);
        }
    }

    return depsResolved as { [k: string]: T };
}

/**
 * Search for a templates dependencies
 * @param templates
 * @param depsResolved Deps resolution cache
 * @returns Map with deps for each template
 */
export function templatesDepsMap(
    templates: ITemplates,
    depsResolved: { [k: string]: string[] } = {},
): { [k: string]: string[] } {
    return runResolveDeps(
        templates,
        ({ name, plainDeps, depsResolved }) => {
            const deps = new Set([name]);

            for (const plainDep of plainDeps) {
                for (const dep of depsResolved[plainDep]) {
                    deps.add(dep);
                }
            }

            return [...deps];
        },
        depsResolved,
    );
}
