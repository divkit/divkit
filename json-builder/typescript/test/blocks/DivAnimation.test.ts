import { divCard, DivText, IDivData, reference, rewriteRefs, templateHelper } from '../../src';

describe('DivAnimation tests', (): void => {
    it('should create card with templates', (): void => {
        const templates = {
            div: new DivText({
                text: 'Some text',
                action_animation: {
                    name: 'set',
                    items: [
                        {
                            name: 'fade',
                            start_value: reference('fade_start_value'),
                            end_value: 1,
                        },
                        {
                            name: 'scale',
                            start_delay: reference('scale_start_delay'),
                            start_value: reference('scale_start_value'),
                            end_value: 1,
                        },
                    ],
                },
            }),
        };

        const divData: IDivData = {
            log_id: 'id',
            states: [
                {
                    state_id: 1,
                    div: templateHelper(templates).div({
                        fade_start_value: 0.5,
                        scale_start_delay: 250,
                        scale_start_value: 0.0,
                    }),
                },
            ],
        };

        expect(divCard(rewriteRefs(templates), divData)).toMatchSnapshot();
    });
});
