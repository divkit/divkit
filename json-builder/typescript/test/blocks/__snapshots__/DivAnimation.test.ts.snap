// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DivAnimation tests should create card with templates 1`] = `
{
  "templates": {
    "div": {
      "type": "text",
      "action_animation": {
        "name": "set",
        "items": [
          {
            "name": "fade",
            "end_value": 1,
            "$start_value": "fade_start_value"
          },
          {
            "name": "scale",
            "end_value": 1,
            "$start_delay": "scale_start_delay",
            "$start_value": "scale_start_value"
          }
        ]
      },
      "text": "Some text"
    }
  },
  "card": {
    "log_id": "id",
    "states": [
      {
        "state_id": 1,
        "div": {
          "type": "div",
          "fade_start_value": 0.5,
          "scale_start_delay": 250,
          "scale_start_value": 0
        }
      }
    ]
  }
}
`;
